### ISSUE TITLE
PyOpenCl

### ISSUE THREAD

**Issue by DjoleStepa** [chaquo/chaquopy#367]
Hi, I try to install PyOpenCl in Andriod Studio, but I have this error.

File "C:\Users\sdjordje\AppData\Local\Temp\easy_install-pu9wdfp0\numpy-1.19.1\tools\cythonize.py", line 59, in process_pyx
        from Cython.Compiler.Version import version as cython_version
    ModuleNotFoundError: No module named 'Cython'.

Install cython and try to install numpy-1.19.1, but same problem. Can you help me?

**Comment by mhsmith** [comment]
Thanks for the request. PyOpenCL isn't currently a priority, but if anyone else wants it too, please click the thumbs up button above, and subscribe to this issue for updates.

As for NumPy, I'm not sure why you're trying to install version 1.19.1, as the current version of PyOpenCL has no version number requirement. Chaquopy currently provides NumPy version 1.17.4: if you need a newer version than that, please explain why.

**Comment by DjoleStepa** [comment]
If I understand, PyOpenCl not available now?

I want NumPy because in error, version of NumPy was 1.19.1. But, if PyOpenCl does not work, Numpy 1.17.4 is ok.

Thanks for answer.

**Comment by mhsmith** [comment]
> If I understand, PyOpenCl not available now?

Unfortunately not.

**Comment by mpottinger** [comment]
I would love to have this. I am currently using compute shaders on Android, but PyOpenCL would be an amazing option to have. How difficult is it to get this to work? Is it something I could work on myself?

**Comment by mhsmith** [comment]
It might be tricky because OpenCL isn't one of the standard Android libraries. But [TensorFlow recently announced](https://blog.tensorflow.org/2020/08/faster-mobile-gpu-inference-with-opencl.html) that they were using it anyway by "check[ing] the availability of OpenCL at runtime".

Since recent versions of Android [don't allow you to access non-standard system libraries](https://android-developers.googleblog.com/2016/06/improving-stability-with-private-cc.html), I'm not sure how this works. The TensorFlow post says that "major Android vendors include OpenCL as part of their system library", so maybe that means it's built it into libc or one of the other standard libraries. 

If you'd like to try building PyOpenCL yourself, then you can license our package build tool, which was used to build everything in our [native package repository](https://chaquo.com/pypi-7.0/). The tool runs on Linux, and comes with an optional Dockerfile for easier dependency management and reproducibility. It supports pybind11, Cython, CMake, and many other build systems, and handles a wide variety of cross-compilation issues.

The cost of this tool is $250 -- this is separate to the license for Chaquopy itself. We would provide support for any general Android build issues, but we have no expertise in OpenCL specifically.

If you're interested in this, please [email us](https://chaquo.com/chaquopy/contact/) for more details.

**Comment by mpottinger** [comment]
> The cost of this tool is $250 -- this is separate to the license for Chaquopy itself. We would provide support for any general Android build issues, but we have no expertise in OpenCL specifically.
> 
> If you're interested in this, please [email us](https://chaquo.com/chaquopy/contact/) for more details.

I was able to load the OpenCL library in Chaquopy using ctypes, initial attempts crashed but I may have been initializing the library incorrectly. It did load libOpenCL.so though and attempt to access the gpu, i could see it loaded mali drivers in logcat.

Currently OpenGL Compute shaders are working well for me, but if I find myself really badly needing a feature of OpenCL that OpenGL lacks then I will definitely take up that offer. That or any other native packages I come across.

**Comment by mhsmith** [comment]
> I was able to load the OpenCL library in Chaquopy using ctypes, initial attempts crashed but I may have been initializing the library incorrectly. It did load libOpenCL.so though

Interesting, thanks. How exactly did you load it: was it as simple as `ctypes.CDLL("libOpenCL.so")`?

**Comment by mpottinger** [comment]
> 
> Interesting, thanks. How exactly did you load it: was it as simple as `ctypes.CDLL("libOpenCL.so")`?

On my Huawei P30 Pro yes it was as simple as that. Hopefully Tensorflow and/or Pytorch get better OpenCL support. So far the tensorflow-gpu package did not seem to actually use the GPU in Chaquopy when I tried it.

**Comment by mhsmith** [comment]
> So far the tensorflow-gpu package did not seem to actually use the GPU

That's correct: since TensorFlow 2.0, the `tensorflow` and `tensorflow-gpu` packages are [actually the same](https://github.com/tensorflow/tensorflow/issues/39581). They only support GPU acceleration via CUDA, which won't work on Android.

As well as OpenCL, the Java TensorFlow Lite library also supports acceleration on Android via OpenGL and NNAPI, but none of these are supported by its Python API yet. Hopefully this will improve in the future, as there also have been some [requests for OpenGL acceleration on the Raspberry Pi](https://github.com/tensorflow/tensorflow/issues/43208).

> Is your tensorflow-gpu package built from this? The SYSCL port of Tensorflow

No, we have our own patches against the main TensorFlow repository.
